parse/EffectPythonParser.cpp

```diff
@@ -0,0 +1,172 @@
+#include "EffectPythonParser.h"
+
+#include <boost/python/extract.hpp>
+#include <boost/python/raw_function.hpp>
+#include <boost/python/stl_iterator.hpp>
+
+#include "../universe/Effects.h"
+#include "../universe/Enums.h"
+
+#include "EnumPythonParser.h"
+#include "ValueRefPythonParser.h"
+
+namespace py = boost::python;
+
+namespace {
+    effect_wrapper insert_generate_sit_rep_message_(const py::tuple& args, const py::dict& kw) {
+        auto message = py::extract<std::string>(kw["message"])();
+        auto label = py::extract<std::string>(kw["label"])();
+
+        bool stringtable_lookup = true;
+        if (kw.has_key("NoStringtableLookup")) {
+            stringtable_lookup = false;
+        }
+
+        std::string icon;
+        if (kw.has_key("icon")) {
+            icon = py::extract<std::string>(kw["icon"])();
+        }
+
+        std::vector<std::pair<std::string, std::unique_ptr<ValueRef::ValueRef<std::string>>>> parameters;
+        if (kw.has_key("parameters")) {
+            py::dict param_args = py::extract<py::dict>(kw["parameters"]);
+            py::stl_input_iterator<std::string> p_begin(param_args.keys()), p_end;
+            for (auto it = p_begin; it != p_end; ++it) {
+                if (auto p_arg = py::extract<value_ref_wrapper<std::string>>(param_args[*it]); p_arg.check()) {
+                    parameters.emplace_back(*it, ValueRef::CloneUnique(p_arg().value_ref));
+                } else if (auto p_arg_double = py::extract<value_ref_wrapper<double>>(param_args[*it]); p_arg_double.check()) {
+                    parameters.emplace_back(*it, std::make_unique<ValueRef::StringCast<double>>(ValueRef::CloneUnique(p_arg_double().value_ref)));
+                } else if (auto p_arg_int = py::extract<value_ref_wrapper<int>>(param_args[*it]); p_arg_int.check()) {
+                    parameters.emplace_back(*it, std::make_unique<ValueRef::StringCast<int>>(ValueRef::CloneUnique(p_arg_int().value_ref)));
+                } else {
+                    parameters.emplace_back(*it, std::make_unique<ValueRef::Constant<std::string>>(py::extract<std::string>(param_args[*it])));
+                }
+            }
+        }
+
+        std::unique_ptr<ValueRef::ValueRef<int>> empire;
+        if (kw.has_key("empire")) {
+            auto empire_args = py::extract<value_ref_wrapper<int>>(kw["empire"]);
```

Comment:
there are probably cases there are unnecessary copies being made of whatever is returned from `py::extract`, which I think can return a const reference. not a high priority to fix, though.